elastic/kibana
# [Metrics UI] Remove TSVB dependency from Metrics Explorer APIs

## Summary

This PR removes the TSVB dependency from Metrics Explorer and introduces a new library called "Metrics API".  The goal of this library is to create a replacement backend for both the Metrics Explorer, Snapshot APIs, and alert executors. This new library supports the following features:

- Multiple group bys
- Multiple metrics defined using ES native aggregations
- Optional alignment of the timeseries to the end of the time range. By default ES aligns the buckets to the beginning of the time range
- Drop last bucket, similar to how TSVB works
- Auto bucketing with a floor, this uses the same syntax `>=1m` as TSVB
- Paginate through groupings
- Everything is Typescript
- Doesn't depend on Metric UI's source or the KIbana request context


Here is an example of how to use this new service.

```typescript
import { query } from 'server/lib/metrics';
import { MetricsAPIRequest, MetricsAPIResponse } from 'common/http_api';
import { createSearchClient } from 'server/lib/create_search_client';

const options: MetricsAPIRequest = {
  timerange: {
    field: '@timestamp',
    from: new Date('2020-01-01T00:00:00Z').getTime(),
    to: new Date('2020-01-01T01:00:00Z').getTime(),
    interval: '>=1m',
  },
  limit: 9,
  dropLastBucket: true,
  indexPattern: 'metrics-*',
  metrics: [
    { id: 'metric_0', aggregations: { metric_0: { avg: { field: 'system.cpu.user.pct' } } } },
  ],
};

const client = createSearchClient(requestContext, framework);
const response: MetricsAPIResponse = await query(client, options);
```

One of the goals of this new service is to provide a consistent way to query time series metrics from anywhere with in the Metrics UI app. The search client function can be swapped out with a function (with the same signature, `ESSearchClient`) that makes calls using a search strategy.  

I plan to turn the Metrics API into a Search Strategy in a separate PR which will make using this library even more ubiquitous. I'm also going to submit a follow up PR to use this API for the backend for the Snapshot API. 

### Testing this PR

The goal of this PR is to go unnoticed by the user. The only change to the integration test was the ID field when the user didn't specify any group by's. Before this change it returned "Everything" as the ID, now it returns `*` and we display the "Everything" title only in the UI when it matches this value.

- Ingest data from Metricbeat running on your localhost
- Open Metrics Explorer on Master and create a visualization
- Check out this new PR, the visualization and data should be identical.



### Checklist

Delete any items that are not applicable to this PR.
